lib/streamlit/server/server.py

```diff
@@ -396,8 +414,52 @@ def _set_state(self, new_state):
         self._state = new_state
 
     @property
-    def is_ready_for_browser_connection(self):
-        return self._state not in (State.INITIAL, State.STOPPING, State.STOPPED)
+    def is_ready_for_browser_connection(self) -> Tuple[bool, str]:
+        if self._state not in (State.INITIAL, State.STOPPING, State.STOPPED):
+            return True, "ok"
+
+        return False, "unavailable"
+
+    """Load and execute the app's script to verify it runs without an error.
+
+    Returns
+    -------
+    [True, "ok"] if the script completes without error, or [False, err_msg]
+    if the script raises an exception.
+    """
+
+    def does_script_run_without_error(self) -> Tuple[bool, str]:
+        if self._script_loading_session_id is None:
+            session = ReportSession(
+                ioloop=self._ioloop,
+                script_path=self._script_path,
+                command_line=self._command_line,
+                uploaded_file_manager=self._uploaded_file_mgr,
+            )
+            self._script_loading_session_id = session.id
+            self._session_info_by_id[self._script_loading_session_id] = SessionInfo(
+                None, session
+            )
+
+        session_info = self._session_info_by_id[self._script_loading_session_id].session
+
+        session_info.session_state.clear_state()
+        session_info.request_rerun(None)
+
+        now = time.perf_counter()
+        while (
+            SCRIPT_RUN_WITHOUT_ERRORS_KEY not in session_info.session_state
+            and (time.perf_counter() - now) < SCRIPT_RUN_CHECK_TIMEOUT
+        ):
+            time.sleep(0.1)
+
+        if SCRIPT_RUN_WITHOUT_ERRORS_KEY not in session_info.session_state:
+            return False, "timeout"
```

Comment:
This function will cause the server to block entirely while the script-health-check is running. Tornado is a [single-threaded web server](https://www.tornadoweb.org/en/stable/guide/async.html), so response handlers need to return very quickly _or_ they need to do their work asynchronously.

For example, if you launch streamlit with this test app:
```python
import streamlit as st

st.write("starting")
time.sleep(55)
st.write("done")
```

and then access the `script-check` endpoint, your Streamlit app will be unavailable for the entire 55 seconds. (Like, you won't even be able to connect to the frontend because you'll quickly get a timeout error.)